Improve QSorted constructor

gassmoeller opened the pull request:
When profiling an ASPECT model I noticed that we spend significant time (~15 % of total runtime in my biased test case) in the constructor of QSorted. This puzzled me, and I will open another PR that suggests to reduce the number of times QSorted is constructed. As a first step this PR optimizes the constructor itself. According to callgrind this change should reduce the number of CPU cycles by around 15%, which I could confirm for my 2D test case (a total model runtime reduction of 2-4% or so in debug and even more in release mode). Since this change should be not critical, I open this as a separate PR.
Let me know if you see other possible improvements, or problems.

bangerth commented:
/run-tests

bangerth commented on include/deal.II/base/quadrature_lib.h:
```diff
@@ -438,13 +438,13 @@ class QSorted : public Quadrature<dim>
-  bool operator()(const std::pair<double, Point<dim> > &a,
-                  const std::pair<double, Point<dim> > &b);
+  bool operator()(const std::size_t &a,
+                  const std::size_t &b) const;
```
Can you explain what `a` and `b` refer to here? Indices into an array whose elements you want to compare?

davydden replied:
Please reflect it in the description of the function.

davydden replied:
maybe mentioning `std::sort` as well, so that it's clear why this function is here.

bangerth replied:
Yes, I actually don't care about this operator at all. It's `public`, and its name gives away nothing. We ought to make this a regular function and use `std::bind` on it.

bangerth replied:
(But it was there before, and so I didn't feel like requesting this.)

gassmoeller replied:
I introduced a new private function `compare_weights` instead. What should happen to the operator? I doubt anybody used it (nothing you can externally use it for really), so should I just remove it? Or deprecate it first?

bangerth replied:
Just remove it.

bangerth commented on source/base/quadrature_lib.cc:
```diff
@@ -932,27 +932,26 @@ QGaussOneOverR<2>::QGaussOneOverR(const unsigned int n,
-                                                quad.point(i)));
-  sort(wp.begin(), wp.end(), *this);
+  std::vector<std::size_t> permutation(quad.size());
+  std::iota(permutation.begin(), permutation.end(), 0);
```
Is `std::iota` already C++98?

gassmoeller replied:
Good point, its C++11, I used a simple loop instead.

bangerth commented:
Does the change with `operator()` actually make any kind of difference? I would think that passing a reference or an integer is about the same. The lack of a reference on the constructor is of course a major problem, and one that I'm sure cost a lot of time.

davydden commented on include/deal.II/base/quadrature_lib.h:
```diff
@@ -438,13 +438,13 @@ class QSorted : public Quadrature<dim>
   /**
    * The constructor takes an arbitrary quadrature formula.
```
`formula.` -> `formula @p quad.`

gassmoeller commented:
I did not benchmark every line change, but the reference in the constructor argument and reserving the appropriate space for the old vector of pairs made up about half of yesterdays improvement. The other half came from sorting the vector of indices and directly copying the data to its new location instead of allocating and sorting additional memory for the pairs of weights and points.

But now that I addressed the comments I noticed another huge improvement in performance (another 75% reduction compared to the initial PR version). My guess: `std::sort` takes either function pointers or function objects as comparison functions and we passed it `*this`, which apparently made copies of the current Quadrature object inside of `std::sort`. Now that I pass a pointer to a function there are way less calls to the copy constructor of Quadrature (and also the destructor), therefore saving a lot of memory allocation / deallocations.
Independent of the discussion in #2990 and #2896 this PR should improve runtimes significantly for anybody using QSorted.

bangerth commented:
Yes, no doubt. Thanks for looking into this -- this is the sort of thing one would never find without actual benchmarking evidence!
